include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,103 @@
+  // The stream has been reset.
+  Reset,
+  // The stream has gRPC errors.
+  // TODO(htuch): Consider providing more granular gRPC failure information (e.g. status code,
+  // framing error, resource exhaustion, authentication failure).
+  GrpcError,
+};
+
+/**
+ * An in-flight gRPC stream.
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   * @param end_stream supplies whether this is the last message.
+   */
+  virtual void sendMessage(const RequestType& request, bool end_stream) PURE;
+
+  /***
+   * Reset the stream.
+   */
+  virtual void reset() PURE;
+};
+
+/**
+ * Notifies caller of async gRPC stream status.
+ * Note the gRPC stream is full-duplex, even if the local to remote stream has been ended by
+ * Stream.sendMessage with end_stream=true, AsyncClientCallbacks can continue to receive events
+ * until the remote to local stream is closed, and vice versa.
+ * TODO(htuch): Add support for metadata.
+ */
+template <class ResponseType> class AsyncClientCallbacks {
+public:
+  virtual ~AsyncClientCallbacks() {}
+
+  /**
+   * Called when an async gRPC message is received.
+   * @param response the gRPC message.
+   */
+  virtual void onMessage(const ResponseType& message) PURE;
+
+  /**
+   * Called when an async gRPC stream is remotely closed with successful status.
+   * @param metadata the trailing metadata.
+   */
+  virtual void onRemoteClose() PURE;
+
+  /**
+   * Called when the async gRPC stream fails. The only valid action at this point is to perform a
+   * reset() on the stream.
+   * @param reason the gRPC FailureReason.
+   */
+  virtual void onFailure(AsyncClientFailureReason reason) PURE;
+};
+
+/**
+ * Supports sending gRPC requests and receiving responses asynchronously. This can be used to
+ * implement either plain gRPC or streaming gRPC calls.
+ */
+template <class RequestType, class ResponseType> class AsyncClient {
+public:
+  virtual ~AsyncClient() {}
+
+  /**
+   * Start a gRPC stream asynchronously.
+   * TODO(htuch): Add support for metadata.
+   * @param service_full_name gRPC service name.
+   * @param method_name gRPC method name.
+   * @param end_stream supplies whether this is a headers-only request.
+   * @param callbacks the callbacks to be notified of stream status.
+   * @param timeout supplies the stream timeout, measured since when the frame with end_stream
+   *        flag is sent until when the first frame is received.
+   * @return a stream handle or nullptr if no stream could be started. NOTE: In this case
+   *         onFailure() has already been called inline. The client owns the stream and
+   *         the handle can be used to send more messages or close the stream.
+   */
+  virtual AsyncClientStream<RequestType>*
+  start(const std::string& service_full_name, const std::string& method_name, bool end_stream,
```

Comment:
Consider take a protobuf descriptor pointer `const google::protobuf::MethodDescriptor*` here instead of service name and method name, it should be easy to get from generated pb.h and it contains more information you might want to use.